java/ql/src/semmle/code/java/frameworks/javase/URI.qll

```diff
@@ -0,0 +1,53 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+
+/* Any expresion or call which returns a new URI.*/
+abstract class URICreation extends Top {
+  /*
+   * Returns the host of the newly created URI.
+   *  In the case where the host is specified separately, this returns only the host.
+   *  In the case where the uri is parsed from an input string,
+   *  such as in `URI(`http://foo.com/mypath')`,
+   *  this returns the entire argument passed i.e. `http://foo.com/mypath'.
+   */
+
+  abstract Expr hostArg();
+}
+
+/* An URI constructor expression */
+class URIConstructor extends ClassInstanceExpr, URICreation {
+  URIConstructor() { this.getConstructor().getDeclaringType().getQualifiedName() = "java.net.URI" }
+
+  override Expr hostArg() {
+    // URI​(String str)
+    if this.getNumArgument() = 1
+    then result = this.getArgument(0)
+    else
+      // URI​(String scheme, String ssp, String fragment) 	
+      if this.getNumArgument() = 3
+      then result = this.getArgument(1)
+      else
+        // URI​(String scheme, String userInfo, String host, int port, String path, String query,
+        //    String fragment) 	
+        if this.getNumArgument() = 7
+        then result = this.getArgument(2)
+        else
+          // URI​(String scheme, String host, String path, String fragment) 	
+          if this.getNumArgument() = 4
+          then result = this.getArgument(1)
+          else
+            // URI​(String scheme, String authority, String path, String query, String fragment)
+            if this.getNumArgument() = 5
+            then result = this.getArgument(1)
+            else none()
```

Comment:
Could this be re-written as a type hierarchy instead? It may not be possible.

Reply:
I don't think a type heirarchy would be a good choice here. I understand that this is a bad coding practice but I have no clue how to eliminate this from my code. @aschackmull Any thoughts on this?

Reply:
Couldn't something like this work?
```ql
result = this.getArgument(0) and this.getNumArgument() = 1
or
result = this.getArgument(1) and this.getNumArgument() = 3
or
result = this.getArgument(2) and this.getNumArgument() = 7
or
result = this.getArgument(1) and this.getNumArgument() = 4
or
result = this.getArgument(1) and this.getNumArgument() = 5
```
Same for `URL` below.

Reply:
This may work for URI but not for URL as there are multiple functions with the same no of args but different signatures. This looks much better though. I will change my code to something similar to what you suggest.

Reply:
> You can also shorten this further by using set literals:

🤯 New syntax! Very cool!